cleanup inheritance structure of CwmsDTO vs CwmsDTOBase

adamkorynta opened the pull request:
classes that were passing in null office ids to CwmsDTO are now using CwmsDTOBase and CwmsDTOBase is made abstract with the default validation logic

Also this updates the validator to use the Java Bean getter instead of using field access and having to modify access. Added caching to reduce performance hit in subsequent calls.

adamkorynta committed:
cleanup inheritance structure of CwmsDTO vs CwmsDTOBase

classes that were passing in null office ids to CwmsDTO are now using CwmsDTOBase and CwmsDTOBase is made abstract with the default validation logic

RyanM-RMA commented on cwms-data-api/src/main/java/cwms/cda/data/dto/Blob.java:
```diff
@@ -58,11 +55,4 @@ public String toString(){
-		super.validateInternal(validator);
-		validator.required(getOfficeId(), "office-id");
-		validator.required(getId(), "id");
-		validator.required(getValue(), "value");
```
Annotate value as required

RyanM-RMA commented on cwms-data-api/src/main/java/cwms/cda/data/dto/Pool.java:
```diff
@@ -3,27 +3,84 @@
 
 @FormattableWith(contentType = Formats.JSONV2, formatter = JsonV2.class)
-public class Pool extends PoolType implements CwmsDTOBase {
+public class Pool extends CwmsDTOBase {
```
Uses tabs, should we format?

MikeNeilson replied:
Yeah, basically if it it isn't a huge imposition fix the formatting.

Basically use judgement. The Github view is good about hiding whitespace changes so it's generally safe to convert tabs to spaces and remove extra spaces at EOL without making review difficult.

adamkorynta replied:
yeah really easy to do in DTO's like this given there isn't logic to check. Reformatted

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/data/dto/CwmsDTOBase.java:
```diff
@@ -1,12 +1,24 @@
  */
-public interface CwmsDTOBase {
-    public  void validate() throws FieldException; // additional validation logic.
+public abstract class CwmsDTOBase {
```
FYI, interfaces can have default methods now. even in Java 8.

adamkorynta replied:
yeah.... the intention of that feature is to provide ways to modify interfaces without breaking backwards compatibility. Given our codebase is self-contained, the compatibility thing isn't an issue. Additionally, you can't mark default methods as final and we don't want the entire validate() method logic to be overwritten (until we come up with a good exception of course), but we can still provide an extension point through the validateInternal() method. Might be sticking too close to the open-closed principle there.

MikeNeilson replied:
Ah, that makes sense.

MikeNeilson reviewed (commented):
Why abstract vs interface with default? No instance variables are required, just the reference to this with the method would have.

RyanM-RMA commented on cwms-data-api/src/main/java/cwms/cda/data/dto/location/kind/PhysicalStructureChange.java:
```diff
@@ -43,7 +42,7 @@
 import java.util.TreeSet;
 
-public abstract class PhysicalStructureChange<T extends Setting> extends CwmsDTO {
+public abstract class PhysicalStructureChange<T extends Setting> extends CwmsDTOBase {
```
Does the `validateInternal` method here need to be updated?

adamkorynta replied:
yes, fixed

adamkorynta committed:
missed validations from code review

adamkorynta committed:
update to use getter methods from java beans instead of field access

this eliminates the need for synchronization which improves performance
also caches the methods for future use
add some basic performance comparisons to show potential overhead on initial check vs follow-up

adamkorynta committed:
remove redundant tests

adamkorynta committed:
update timeouts for GitHub runs

rma-psmorris commented on cwms-data-api/src/main/java/cwms/cda/data/dto/CwmsDTOValidator.java:
```diff
@@ -29,85 +29,118 @@
 
     private static final FluentLogger LOGGER = FluentLogger.forEnclosingClass();
-
+    private static final Map<Class<?>, List<Method>> REQUIRED_FIELD_GETTERS = new ConcurrentHashMap<>();
```
Did you consider using Java References here to allow for garbage collection?

adamkorynta replied:
yeah, I did but there would be no added benefit. the classloader already caches the Class objects so there is no needed Soft/Weak reference here. Additionally, the Methods are strongly referenced by the Class so reference doesn't help here either.

rma-psmorris commented on cwms-data-api/src/main/java/cwms/cda/data/dto/CwmsDTOValidator.java:
```diff
@@ -29,85 +29,118 @@
-            LOGGER.atWarning().withCause(e).log("Unable to validate required fields are non-null in DTO: " + type);
+        } catch (IntrospectionException e) {
+            LOGGER.atWarning().withCause(e)
+                .log("Unable to validate required field are non-null in DTO: %s", type);
```
"Unable to validate required field are non-null in DTO" --  "required fields are" or "required field is"?

adamkorynta replied:
Addressed in: https://github.com/USACE/cwms-data-api/pull/789

rma-psmorris commented on cwms-data-api/src/main/java/cwms/cda/data/dto/LookupType.java:
```diff
@@ -25,23 +25,23 @@
 @JsonInclude(JsonInclude.Include.NON_NULL)
 @JsonNaming(PropertyNamingStrategies.KebabCaseStrategy.class)
-public final class LookupType implements CwmsDTOBase {
+public final class LookupType extends CwmsDTOBase {
```
This could be a CwmsDTO?

adamkorynta replied:
Updated LookupType and Property: https://github.com/USACE/cwms-data-api/pull/788
VerticalDatumInfo is in a similar boat, but unfortunately uses "office" instead of "office-id" so I'm leaving it for backwards compatibility.

rma-psmorris commented on cwms-data-api/src/main/java/cwms/cda/data/dto/Pool.java:
```diff
@@ -1,246 +1,249 @@
-
+public class Pool extends CwmsDTOBase {
+
+    private final PoolNameType poolName;
```
Recommend this dependency be changed. No longer use PoolNameType and instead add its attributes to Pool. If not possible, then introduce a new PoolId class that has the office and pool name variables.

adamkorynta replied:
Added as an item to accomplish in: https://github.com/USACE/cwms-data-api/issues/786

rma-psmorris commented on cwms-data-api/src/main/java/cwms/cda/data/dto/TimeZone.java:
```diff
@@ -20,30 +21,26 @@
-public final class TimeZone extends CwmsDTO
-{
-	private String timeZone;
+public final class TimeZone extends CwmsDTOBase {
```
Recommend changing this class name to be TimeZoneName or TimeZoneId to clarify it as not having TimeZone business rules and separate it from the java class sharing the same name.

adamkorynta replied:
Renamed to `TimeZoneId` in: https://github.com/USACE/cwms-data-api/pull/787

rma-psmorris commented on cwms-data-api/src/main/java/cwms/cda/data/dto/Unit.java:
```diff
@@ -23,7 +23,7 @@
 @FormattableWith(contentType = Formats.XMLV2, formatter = XMLv2.class, aliases = {Formats.XML})
 @JsonNaming(PropertyNamingStrategies.KebabCaseStrategy.class)
-public final class Unit extends CwmsDTO
+public final class Unit extends CwmsDTOBase
```
I'm not certain if this class was implemented with knowledge of the database notion of DisplayUnits? See source for AV_DISPLAY_UNIT which brings in office specific parameter units

adamkorynta replied:
@RyanM-RMA - I'll let you chime in here.

RyanM-RMA replied:
It was not implemented with that knowledge.  The table and view it's using, AT_UNIT_ALIAS, and AV_UNIT, both contain office IDs, so we could make a ticket to update them to include those office ids.  Seems a bit out of scope for this work though.

RyanM-RMA replied:
Actually, looking at the original output from this (it was using CWMS_CAT.RETRIEVE_UNITS_F before), that output didn't include the office id, which is probably why it was ignored in the changes I made.

rma-psmorris replied:
The importance of office ID as used in display vs the non-office ID CWMS defaults is that the unit will be the one that the district has determined to be their default. @RyanM-RMA - How is that out of scope?

RyanM-RMA replied:
I say out of scope because it's related to recent maintenance ticket work, not our fixed cost task.

RyanM-RMA replied:
This is the ticket for that work.
https://github.com/USACE/cwms-data-api/issues/498

rma-psmorris reviewed (commented):
Post-merge feedback.
